File: fig2_rmse_decomposed.py
#!/usr/bin/env python3
# ============================================================
# fig2_rmse_decomposed.py
#
# Generates Figure 2: RMSE decomposed by true-association status
# (true-nonzero signal pairs vs true-zero null pairs)
#
# UPDATE SCENARIO 2 VALUES when new simulation run completes.
# Scenarios 1 and 3 are final (n=100 replicates).
#
# Run: python fig2_rmse_decomposed.py
# Output: fig2_rmse_decomposed.png (300 DPI)
# ============================================================

import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import numpy as np

# ── Global style ─────────────────────────────────────────────────────────
plt.rcParams.update({
    'font.family':       'DejaVu Sans',
    'font.size':         11,
    'axes.titlesize':    12,
    'axes.labelsize':    11,
    'xtick.labelsize':   10,
    'ytick.labelsize':   10,
    'savefig.dpi':       300,
    'savefig.bbox':      'tight',
    'axes.spines.top':   False,
    'axes.spines.right': False,
})

# ── Colours (colour-blind friendly) ─────────────────────────────────────
COL = {
    'full':         '#0072B2',   # blue
    'no_network':   '#E69F00',   # orange
    'no_horseshoe': '#CC79A7',   # pink
}
LAB = {
    'full':         'Full model',
    'no_network':   'No-network',
    'no_horseshoe': 'No-horseshoe',
}
MODELS = ['full', 'no_network', 'no_horseshoe']

# ── Data ─────────────────────────────────────────────────────────────────
# Each entry: {'signal': RMSE on true nonzero pairs,
#              'null':   RMSE on true zero pairs}
#
# UPDATE sc2 values below when Scenario 2 rerun completes.
# Replace the four numbers marked with  <-- UPDATE

data = {
    # ── Scenario 1: Sparse Independent (FINAL, n=100) ────────────────
    'sc1': {
        'full':         {'signal': 0.562, 'null': 0.066},
        'no_network':   {'signal': 0.263, 'null': 0.102},
        'no_horseshoe': {'signal': 0.415, 'null': 0.166},
    },
    # ── Scenario 2: Network Structured (UPDATE WHEN RERUN COMPLETES) ─
    'sc2': {
        'full':         {'signal': 0.586, 'null': 0.067},   # <-- UPDATE
        'no_network':   {'signal': 0.263, 'null': 0.099},   # <-- UPDATE
        'no_horseshoe': {'signal': 0.418, 'null': 0.166},   # <-- UPDATE
    },
    # ── Scenario 3: Dense Weak Noise (FINAL, n=100) ───────────────────
    'sc3': {
        'full':         {'signal': 0.562, 'null': 0.073},
        'no_network':   {'signal': 0.259, 'null': 0.104},
        'no_horseshoe': {'signal': 0.411, 'null': 0.167},
    },
}

SC_LABELS = {
    'sc1': 'Sc.1: Sparse\nIndependent',
    'sc2': 'Sc.2: Network\nStructured',
    'sc3': 'Sc.3: Dense\nWeak Noise',
}

# ── Plot ─────────────────────────────────────────────────────────────────
fig, axes = plt.subplots(1, 3, figsize=(13, 5))
fig.suptitle(
    'Figure 2  |  RMSE Decomposed by True-Association Status (n = 100 replicates)',
    fontweight='bold', y=1.01
)

x     = np.arange(len(MODELS))
width = 0.35

for col, sc in enumerate(['sc1', 'sc2', 'sc3']):
    ax = axes[col]

    sig_vals  = [data[sc][m]['signal'] for m in MODELS]
    null_vals = [data[sc][m]['null']   for m in MODELS]
    colors    = [COL[m] for m in MODELS]

    # Solid bars = true nonzero (signal) pairs
    bars_sig = ax.bar(x - width / 2, sig_vals, width,
                      color=colors, alpha=0.87)

    # Hatched bars = true zero (null) pairs
    bars_null = ax.bar(x + width / 2, null_vals, width,
                       color=colors, alpha=0.40, hatch='//')

    # Annotate signal bars (bold)
    for bar, v in zip(bars_sig, sig_vals):
        ax.text(bar.get_x() + bar.get_width() / 2,
                bar.get_height() + 0.005,
                f'{v:.3f}',
                ha='center', va='bottom',
                fontsize=8.5, fontweight='bold')

    # Annotate null bars (regular weight)
    for bar, v in zip(bars_null, null_vals):
        ax.text(bar.get_x() + bar.get_width() / 2,
                bar.get_height() + 0.005,
                f'{v:.3f}',
                ha='center', va='bottom',
                fontsize=8.5)

    ax.set_title(SC_LABELS[sc], fontweight='bold', pad=8)
    ax.set_xticks(x)
    ax.set_xticklabels(['Full\nmodel', 'No-\nnetwork', 'No-\nhorseshoe'],
                       fontsize=9.5)
    ax.set_ylabel('RMSE' if col == 0 else '')
    ax.set_ylim(0, 0.70)
    ax.grid(axis='y', alpha=0.25, zorder=0)

# ── Legends ──────────────────────────────────────────────────────────────
# Legend 1: solid vs hatched (subset type)
from matplotlib.patches import Patch
subset_legend = [
    Patch(facecolor='grey', alpha=0.87,
          label='True nonzero pairs (3.3% of all pairs)'),
    Patch(facecolor='grey', alpha=0.40, hatch='//',
          label='True zero pairs (96.7% of all pairs)'),
]
leg1 = fig.legend(handles=subset_legend,
                  loc='lower center', ncol=2, frameon=False,
                  bbox_to_anchor=(0.5, -0.10), fontsize=10)

# Legend 2: model colours
model_handles = [mpatches.Patch(color=COL[m], label=LAB[m], alpha=0.87)
                 for m in MODELS]
fig.legend(handles=model_handles,
           loc='lower center', ncol=3, frameon=False,
           bbox_to_anchor=(0.5, -0.22), fontsize=10)

fig.add_artist(leg1)   # keep both legends visible

# ── Save ─────────────────────────────────────────────────────────────────
plt.tight_layout(rect=[0, 0.12, 1, 1])
fig.savefig('fig2_rmse_decomposed.png', dpi=300, bbox_inches='tight')
plt.close()

print("Saved: fig2_rmse_decomposed.png")
print()
print("To update Scenario 2:")
print("  Edit the three lines marked  # <-- UPDATE  in the data dict above,")
print("  then re-run:  python fig2_rmse_decomposed.py")

File: fig2_rmse_decomposed.R
# ============================================================
# fig2_rmse_decomposed.R
#
# Generates Figure: RMSE decomposed by true-association status
# (true-nonzero signal pairs vs true-zero null pairs)
#
# UPDATE SCENARIO 2 VALUES HERE when new simulation run completes.
# All other values are final (n=100 replicates).
# ============================================================

library(ggplot2)
library(dplyr)
library(tidyr)

# ── Data: update sc2 values when Scenario 2 rerun completes ─────────────
rmse_data <- tribble(
  ~scenario,              ~model,         ~signal, ~null,
  # Scenario 1 (FINAL - n=100)
  "Sc.1: Sparse\nIndependent", "Full model",    0.562,  0.066,
  "Sc.1: Sparse\nIndependent", "No-network",    0.263,  0.102,
  "Sc.1: Sparse\nIndependent", "No-horseshoe",  0.415,  0.166,
  # Scenario 2 (UPDATE THESE when new run completes)
  "Sc.2: Network\nStructured", "Full model",    0.586,  0.067,
  "Sc.2: Network\nStructured", "No-network",    0.263,  0.099,
  "Sc.2: Network\nStructured", "No-horseshoe",  0.418,  0.166,
  # Scenario 3 (FINAL - n=100)
  "Sc.3: Dense\nWeak Noise",   "Full model",    0.562,  0.073,
  "Sc.3: Dense\nWeak Noise",   "No-network",    0.259,  0.104,
  "Sc.3: Dense\nWeak Noise",   "No-horseshoe",  0.411,  0.167
)

# ── Reshape to long format ───────────────────────────────────────────────
rmse_long <- rmse_data %>%
  pivot_longer(c(signal, null),
               names_to  = "subset",
               values_to = "rmse") %>%
  mutate(
    scenario = factor(scenario,
                      levels = c("Sc.1: Sparse\nIndependent",
                                 "Sc.2: Network\nStructured",
                                 "Sc.3: Dense\nWeak Noise")),
    model    = factor(model,
                      levels = c("Full model", "No-network", "No-horseshoe")),
    subset   = factor(subset,
                      levels = c("signal", "null"),
                      labels = c("True nonzero pairs\n(3.3% of all pairs)",
                                 "True zero pairs\n(96.7% of all pairs)"))
  )

# ── Colours (colour-blind friendly) ─────────────────────────────────────
model_colours <- c(
  "Full model"    = "#0072B2",
  "No-network"    = "#E69F00",
  "No-horseshoe"  = "#CC79A7"
)

# ── Plot ─────────────────────────────────────────────────────────────────
p <- ggplot(rmse_long,
            aes(x      = model,
                y      = rmse,
                fill   = model,
                alpha  = subset,
                pattern = subset)) +

  # Bars -- solid for signal, hatched for null
  geom_col(position = position_dodge(width = 0.75),
           width    = 0.35,
           colour   = "white", linewidth = 0.3) +

  # Value labels
  geom_text(aes(label    = sprintf("%.3f", rmse),
                fontface = ifelse(subset ==
                                    "True nonzero pairs\n(3.3% of all pairs)",
                                  "bold", "plain")),
            position = position_dodge(width = 0.75),
            vjust    = -0.4,
            size     = 2.8) +

  facet_wrap(~scenario, nrow = 1) +

  scale_fill_manual(values = model_colours, name = NULL) +
  scale_alpha_manual(
    values = c("True nonzero pairs\n(3.3% of all pairs)" = 0.87,
               "True zero pairs\n(96.7% of all pairs)"   = 0.40),
    name   = NULL
  ) +

  scale_y_continuous(limits = c(0, 0.72),
                     breaks = seq(0, 0.7, by = 0.1),
                     expand = expansion(mult = c(0, 0.05))) +

  labs(
    title = "Figure 2  |  RMSE Decomposed by True-Association Status (n = 100 replicates)",
    x     = NULL,
    y     = "RMSE"
  ) +

  theme_bw(base_size = 11) +
  theme(
    legend.position   = "bottom",
    legend.box        = "vertical",
    legend.key.size   = unit(0.5, "cm"),
    panel.grid.minor  = element_blank(),
    panel.grid.major.x = element_blank(),
    strip.background  = element_rect(fill = "grey92"),
    strip.text        = element_text(face = "bold", size = 10),
    plot.title        = element_text(face = "bold", size = 11),
    axis.text.x       = element_blank(),
    axis.ticks.x      = element_blank()
  ) +

  # Separate legends for model colour and subset pattern
  guides(
    fill  = guide_legend(order = 2, nrow = 1,
                         override.aes = list(alpha = 0.87)),
    alpha = guide_legend(order = 1, nrow = 1,
                         override.aes = list(fill = "grey50"))
  )

# ── Save at 300 DPI ──────────────────────────────────────────────────────
ggsave("fig2_rmse_decomposed.png",
       plot   = p,
       width  = 13,
       height = 5,
       dpi    = 300,
       units  = "in",
       bg     = "white")

cat("Saved: fig2_rmse_decomposed.png\n")
cat("\nTo update Scenario 2: edit the three sc2 rows in rmse_data above\n")
cat("and re-run the script. Everything else is automatic.\n")
